File: 08.Sort/786.第k个数.heap.java
// https://www.acwing.com/problem/content/description/788/
import java.io.*;
import java.util.*;

public class Main {
  static int[] A = new int[100001];
  public static void main(String args[]) throws Exception {
    Scanner sc = new Scanner(System.in);
    int n = sc.nextInt();
    int k = sc.nextInt();

    for (int i = 0; i < n; i++) {
      A[i] = sc.nextInt();
    }
    // 这里用最大堆
    Queue<Integer> Q = new PriorityQueue<>((v1, v2) -> v2 - v1);
    for (int i = 0; i < n; i++) {
      final int x = A[i];
      Q.offer(x);
      while (Q.size() > k) {
        Q.poll();
      }
    }
    System.out.println(Q.peek());
  }
}

File: 08.Sort/786.第k个数.heap.py
# https://www.acwing.com/problem/content/description/788/
import Queue
from Queue import PriorityQueue

s = raw_input().split()
n = int(s[0])
k = int(s[1])
s = raw_input().split()
A = [int(x) for x in s]

Q = PriorityQueue()
for x in A:
  Q.put(-x)
  while Q.qsize() > k:
    Q.get()

print 0-Q.get()


